Social Cohesion - untracked MICalcs

mattstevenson22 opened the pull request.

MattSScott commented on internal/server/Spawner.go:
```diff
@@ -52,29 +47,60 @@ func (s *Server) BikerAgentGenerator(initFunc func(baseBiker *objects.BaseBiker)
 
 func (s *Server) replenishMegaBikes() {
+	// currently not being called as awdi doesnt currently remove the megabike.
+	// if it does, this function needs changing to spawn new megabikes in with the correct governance
 	neededBikes := globals.MegaBikeCount - len(s.megaBikes)
 	for i := 0; i < neededBikes; i++ {
-		s.spawnMegaBike()
+		s.spawnMegaBike(utils.One)
 	}
```
Is this what you want? Not an equal split of all three types? Who defined this?

mattstevenson22 replied:
I'm a bit confused sorry! I did not write this code, but it is called if the awdi destroys a megabike. currently the awdidestroysmegabike parameter or whatever its called is set to false i think, so i didnt worry about this and just wrote in whatever. if it was being called though, this function should be changed so that it can respawn in the bikes that have been destroyed.

MattSScott reviewed (changes requested):
Looks really good, I don't really have any complaints. I'll submit as a "request changes now", just so that you see my comments, but you can accept/dismiss at your behest and ask me to re-review

mattstevenson22 commented on internal/clients/teamSOSA/modules/AgentParameters.go:
```diff
@@ -7,29 +7,12 @@ import (
 type AgentParameters struct {
-	Trustworthiness float64               // intrinsic value for P(conform to action)
-	TrustNetwork    map[uuid.UUID]float64 // mapping of social network to trust score
+	PreferenceForEquality float64               // hardwired value between [0, 1] reflecting the agents personality, where a higher value signifies greater preference for equal and fair distribution.
```
yes, i will make this change

mattstevenson22 commented on internal/clients/teamSOSA/modules/Environment.go:
```diff
@@ -190,97 +196,54 @@ func (e *EnvironmentModule) GetNearestLootboxAwayFromAwdi() uuid.UUID {
 	bikes := e.GetBikes()
 	for bikeId, bike := range bikes {
-		totalSocialCapital := float64(0)
+		totalTrust := float64(0)
```
made this change too

mattstevenson22 commented on internal/clients/teamSOSA/modules/Environment.go:
```diff
@@ -334,19 +297,75 @@ func (e *EnvironmentModule) GetBikerAgents() map[uuid.UUID]objects.IBaseBiker {
+
+// // returns: the biker on your bike with the minimum trust 
+// func (e *EnvironmentModule) GetBikerWithMinTrust(ap *AgentParameters) IDTrustPair {
+// 	fellowBikers := e.GetBikerAgents() // MATT- THIS LINE GETS ALL AGENTS ON BIKES. SO THE FUNCTION FINDS THE BIKER IN WHOLE GAME WITH LOWEST TRUST. BUT IT WAS USED IN DECIDEKICKOUT, SO IT COULD SUGGEST A BIKER NOT ON YOUR BIKE WHICH IS STUPID. 
```
this was not addressed to you "matt", but me "matt" so  i could remember that i wrote it. sorry!!!

mattstevenson22 commented on internal/common/objects/MegaBike.go:
```diff
@@ -136,77 +105,127 @@ func (mb *MegaBike) UpdateOrientation() {
 
-func (mb *MegaBike) ResetCurrentPool() {
-	mb.currentPool = 0
+// removes: an agent from the bike, given its ID
+func (mb *MegaBike) RemoveAgent(bikerId uuid.UUID) {
+	// Create a new slice to store the updated agents
+	var updatedAgents []IBaseBiker
+
+	// Iterate through the agents and copy them to the updatedAgents slice
+	for _, agent := range mb.agents {
+		if agent.GetID() != bikerId {
+			updatedAgents = append(updatedAgents, agent)
+		}
+	}
+
+	// Replace the mb.agents slice with the updatedAgents slice
+	mb.agents = updatedAgents
```
I found it like this and it seemed like a fine way to do it so I just left it

mattstevenson22 replied:
This has been addressed

mattstevenson22 commented on internal/server/Governance.go:
```diff
@@ -6,42 +6,189 @@ import (
-	votes := make(map[uuid.UUID]voting.IdVoteMap, len(agents))
-	voteWeight := make(map[uuid.UUID]float64)
+// returns: uuid of lootbox to aim toward (i.e. direction) for current round from the agents
+func (s *Server) RunDemocraticDirectionDecision(bike objects.IMegaBike) uuid.UUID {
```
okay thanks for feedback, i will try to bring these all together in a cleaner way in the manner you mentioned

mattstevenson22 commented:
I've made some change based on the feedback and pushed them here

MattSScott commented on internal/clients/teamSOSA/agent/actions.go:
```diff
@@ -100,6 +114,45 @@ func (a *AgentSOSA) DecideChangeBike() uuid.UUID {
+		bikeAverageTrust := sum / float64(len(bike.GetAgents()))
+
+		// Next get the regimeTrust
+		regimeTrust := a.Modules.AgentParameters.RegimeTrust[bike.GetGovernance()]
```
Is the `RegimeTrust` not a piece of 'global' information? It's probably better defining a separate module for it on a server level, instead of having every agent record the data themselves.

mattstevenson22 replied:
See comment below

MattSScott commented on internal/clients/teamSOSA/modules/AgentParameters.go:
```diff
@@ -9,6 +10,7 @@ import (
 type AgentParameters struct {
 	PlatonicTendency float64               // hardwired value between [0, 1] reflecting the agents personality, where a higher value signifies greater preference for equal and fair distribution.
 	TrustNetwork    map[uuid.UUID]float64 		// mapping of uuid -> trust score. trust score ranges from 0 to 1
+	RegimeTrust 	map[utils.Governance]float64 // mapping of trust in the regimes
```
Same as above comment - this should be shared information across all agents, so there's no point having them record it on a local level. Consider moving this up into a common module on the server level, and allowing agents to invoke this

mattstevenson22 replied:
See comment below

MattSScott commented on internal/clients/teamSOSA/modules/AgentParameters.go:
```diff
@@ -74,6 +76,24 @@ func (ap *AgentParameters) UpdateTrustValue(agentID uuid.UUID, eventValue float6
+			ap.RegimeTrust[governance] = 0.5
+		} else {
+			if rank == 0 {
+				ap.RegimeTrust[governance] += 0.05
+				ap.RegimeTrust[governance] = clamp(ap.RegimeTrust[governance])
+			} else if rank == 2 {
+				ap.RegimeTrust[governance] -= 0.05
+				ap.RegimeTrust[governance] = clamp(ap.RegimeTrust[governance])
```
Ok, this code makes my previous two comments irrelevant. Did we not decide that the regime score is just the proportion of total resources that they brought in?

MattSScott replied:
Unless this is trying to make the rankings more 'definitive', by shifting best and worst a bit? If so, this can still be handled globally, right?

MattSScott replied:
Because using proportion makes sure that it's constrained within [0, 1]. If you wanted the dynamic-ness, I'd attach a method to the server, and inject it into the agent (Inversion of Control). The agent can then invoke this method on the server (which itself is dynamically changing)

MattSScott commented on internal/common/objects/BaseBiker.go:
```diff
@@ -81,7 +87,7 @@ type BaseBiker struct {
 	gameState                        IGameState // updated by the server at every round
 	roundDecisions					 roundDecisions // agent keeps a track of the decisions it makes each round to message other agents at the end of the round
-
+	developedCohesion				 bool			// has this agent started prioritising trust in agents over regime
```
I thought that we defined this as simply whether the dominant factor in the preference order was the regime or the agents. Isn't it easier to submit a 3-tuple in the preference order? ID, score, and flag as to whether trust dominates? The server can then parse this, and make a running total of the number of bike selections, and the number of times that this variable dominated?

mattstevenson22 replied:
Good idea, this makes more sense that what I've done I think regarding the developed cohesion thing. I'll try and implement something like this

MattSScott commented on internal/common/objects/BaseBiker.go:
```diff
@@ -471,3 +481,23 @@ func (bb *BaseBiker) GetRoundForces() utils.Forces {
+	return bb.developedCohesion
+}
+
+func (bb *BaseBiker) SetDevelopedCohesion(answer bool) {
+	bb.developedCohesion = answer
+}
```
You would only have a setter if an external class is setting this value. Surely the agent can control this itself? I also think that it makes less work for yourself if you implement my recommendation from above

mattstevenson22 replied:
👍

MattSScott commented on internal/server/Server.go:
```diff
@@ -85,8 +91,18 @@ func (s *Server) Start() {
+				numAgentsDevelopedCohesion += 1
+				delete(agentMapCopy, agentID)
+			}
+		}
```
I don't think that the agent could know this. They don't know which bike that they get from their preference order, and the information about which value dominated is lost when you just smash them into a weighted sum. You should create that 3-tuple, and have the server perform this update once the agent is allocated. I.e.:

```
for (bikeID, _, isCohesive) := range preferenceOrder {
    // some logic
    if canAssign and isCohesive...
}
```

mattstevenson22 replied:
👍 👍

MattSScott commented on internal/server/SimLoop.go:
```diff
@@ -67,8 +84,33 @@ func (s *Server) RunSimLoop(rounds int, gameState *SimplifiedGameStateDump, iter
 
+	// experimental
+
+	resourcesGainedPerRegime := make(map[utils.Governance]float64)
+	for _, bike := range s.GetMegaBikes() {
+		resourcesGainedPerRegime[bike.GetGovernance()] = bike.GetCurrentPool() - initialCurrentPoolPerRegime[bike.GetGovernance()]
+	}
```
Not sure I fully get this. Why are we not just taking the current pool?

mattstevenson22 replied:
They start with trusting each regime equally, and adjust their score each iteration based on how the regimes performed that iteration. So this is calculating the bikes current pool at end of iteration - the initial current pool at start of iteration to see how many resources this regime gained this iteration

MattSScott replied:
Ah ok, nice. It becomes a question of whether you want a longitudinal score (i.e., proportion of resources gained in the entire game) or just on a round basis. Maybe the former is a bit more informative?

MattSScott commented on internal/server/SimLoop.go:
```diff
@@ -357,6 +400,139 @@ func (s *Server) SetDestinationBikes() {
+			s.AddAgentToBike(s.GetAgentMap()[chosenReps[0]], bike)
+			bike.SetRepresentatives([]uuid.UUID{chosenReps[0]})
+		}
+	}
```
This is a bit dangerous. Firstly, you've hard-coded the 4. What if there are twice as many bikes? Three times as many? This breaks the scaling.

You also muddy the clarity by smashing all of the reps into a single value, and a single loop. It would be a lot clearer to manually go bike by bike as follows:

1. stratify all bike by regime
2. for all ONE bikes, go one by one and try to assign a representative
3. for all SOME bike, go one by one and try to assign three representatives

You can do this with a simple two-pointer approach. Keep a `repPointer` that increments one-by-one for each successful allocation. If you reach the end of the bikes, or reach the end of id array, then you break

mattstevenson22 replied:
Yeah definitely, I got a bit lazy late last night, hardcoding things is always a bad idea. I'll try to make these changes now

MattSScott commented on internal/server/SimLoop.go:
```diff
@@ -357,6 +400,139 @@ func (s *Server) SetDestinationBikes() {
+	}
+
+	// Step 3: Go through the queued agents, asking them their top bike, then prompting that bike to make an acceptance decision.
+	for i := 0; i < 3; i++ {
```
This loop isn't clear to me, sorry

MattSScott reviewed (changes requested):
Some things there, that I think will help with reproducibility, scaling, and clarity. Let me know if you want some help with implementation

mattstevenson22 commented:
I made plenty of changes addressing your feedback which was very helpful, especially the stuff regarding scaling. I've tried to address this with these changes, as well as doing some general tidying and refactoring into functions to make things easier to read.

MattSScott commented on internal/clients/teamSOSA/modules/AgentParameters.go:
```diff
@@ -108,7 +108,7 @@ func clamp(value float64) float64 {
 func NewAgentParameters() *AgentParameters {
 	return &AgentParameters{
-		PlatonicTendency: rand.Float64(),
+		PlatonicTendency: 0.1,
```
Shouldn't this be accessed from some flag or common initialiser?

Hard-coding 0.1 basically makes them all b*stards :))

mattstevenson22 replied:
Yes, I was experimenting with different distributions and must have left it on 0.1 lmao. I will add some flag for this

MattSScott commented on internal/common/objects/BaseBiker.go:
```diff
@@ -29,7 +29,7 @@ type IBaseBiker interface {
 	DecideForce(direction uuid.UUID)                             // decides: the force the biker is going to pedal with
 	HandleAgentUnalive(id uuid.UUID)                             // decides: how to handle a dead agent
-	DecideBikePreferenceOrder() []uuid.UUID						// decides: the order of the agents preferred bikes
+	DecideBikePreferenceOrder() ([]uuid.UUID, map[uuid.UUID]bool)					// decides: the order of the agents preferred bikes
```
Should return `[]BikePreferenceData`

mattstevenson22 replied:
:thumbs up:

MattSScott commented on internal/server/SimLoop.go:
```diff
@@ -431,58 +470,95 @@ func (s *Server) RunVoluntaryReassociation() {
+		for _, id := range repIDs {
+			s.AddAgentToBike(agentMap[id], bike)
+		}
+		bike.SetRepresentatives(repIDs)
```
Much easier to just add a nested loop of 3 with golang labels

```
someBikeLoop:
for _, bike := range someBikes {
  for range 3 {
    if repPointer == end {break someBikeLoop}
    // same allocation logic 
  }
}
```

MattSScott replied:
You can actually label the outer loop, such that you break from it (and not the innermost)

mattstevenson22 replied:
Don't quite get this sorry, not sure I know what you mean by labels?

MattSScott replied:
Google it haha. The code I wrote above is basically valid. You can manually label loops - for example, I've labelled your outer loop as "someBikeLoop" in the code snippet. This allows you to explicitly say something like `break someBikeLoop`, which breaks from the outer loop instead of the innermost, like in most languages

mattstevenson22 replied:
Okay I think I've implemented this now

MattSScott commented on internal/server/SimLoop.go:
```diff
@@ -431,58 +470,95 @@ func (s *Server) RunVoluntaryReassociation() {
+	// add all the non-rep agents to the queue.
 	for agentID, agent := range agentMap {
-		if slices.Contains(chosenReps, agentID) {
+		if assignedReps[agentID] {
```
Isn't this where you need the `ok`?

`if rep, ok := assignedReps[agentID] ; !ok`

mattstevenson22 replied:
Not sure what you mean by this. I just thought it was a simple case of going through the agent map, and if they are an assigned rep move to the next agent, otherwise add them to the queue?

MattSScott replied:
Yes, but slices.Contains is $O(n)$ lookup

The idiomatic way in golang is what I wrote above; it's the syntactic equivalent of a pythonic `if agentID in assignedReps`

mattstevenson22 replied:
Ok, so something like this:

	for agentID, agent := range agentMap {
		if _, ok := assignedReps[agentID] ; !ok {
			queuedAgents[agentID] = agent
		}
	}

MattSScott commented on internal/server/SimLoop.go:
```diff
@@ -431,58 +470,95 @@ func (s *Server) RunVoluntaryReassociation() {
-					for _, agent := range topBike.GetAgents() {
+					for _, agent := range nextHighestBike.GetAgents() {
 						decisions = append(decisions, agent.DecideJoiningOneAgent(agentNextInLineId))
 					}
```
Much cleaner to have a separate function for this:

`finalDecision := serv.GetManyJoiningDecision(nextHighestBike)`

MattSScott replied:
And the same for the others, if you don't mind

mattstevenson22 replied:
Will do this too, ty

MattSScott reviewed (changes requested):
Getting really close! Still one or two things that seem slightly off, but I reckon we've almost cracked it

mattstevenson22 commented on internal/clients/teamSOSA/agent/actions.go:
```diff
@@ -114,43 +114,42 @@ func (a *AgentSOSA) DecideChangeBike() uuid.UUID {
 
-	return bikePreferenceOrder
-
+	return bikePreferenceOrder, bikeTrustHigherThanRegime
```
👍 I'll get on this now

mattstevenson22 commented:
Made most of these changes and pushed again. Once reviewed I can start adding the stuff about Gini etc

MattSScott commented on internal/clients/teamSOSA/modules/AgentParameters.go:
```diff
@@ -108,7 +108,7 @@ func clamp(value float64) float64 {
 func NewAgentParameters() *AgentParameters {
 	return &AgentParameters{
-		PlatonicTendency: 0.1,
+		PlatonicTendency: rand.Float64(),
```
Sure, but we'll need some way of fixing this for the experimentation. You may find yourself having to pass this in as an argument to the function, as the distribution will likely be set in the server/main.go

This said, we need to make a fresh PR. This one is massively oversaturated. Can you implement this on a fresh `experimentation` branch after?

MattSScott commented on internal/server/SimLoop.go:
```diff
@@ -549,54 +549,18 @@ func (s *Server) ProcessAgentQueue(queuedAgents map[uuid.UUID]objects.IBaseBiker
-			bikePreferenceOrder, bikeTrustHigherThanRegimeTrustMap := agentNextInLine.DecideBikePreferenceOrder()
-			nextHighestBike := s.GetMegaBikes()[bikePreferenceOrder[i]]
+			bikePreferenceOrder := agentNextInLine.DecideBikePreferenceOrder()
+			nextHighestBike := s.GetMegaBikes()[bikePreferenceOrder[i].BikeId]
```
This line of code is used a ton. Could you add a method to the server with something like `GetBikeByID(id)`?

MattSScott reviewed (approved):
Looks really good! I've left two comments, but I don't think they're worth continuing this back-and-forth for. Could you address them in a fresh PR, please?

MattSScott reviewed (changes requested):
Still a bit missing - will reply to comments you're not sure on

mattstevenson22 commented on internal/clients/teamSOSA/agent/messages.go:
```diff
@@ -2,95 +2,219 @@ package agent
+		} else {
+			a.Modules.AgentParameters.UpdateTrustValue(sender.GetID(), modules.NegativeMessage)
+		}
+	} 
```
this has been addressed

MattSScott reviewed (approved):
Nice work :))
